plantmanager-react/src/components/Login/Login.js

```diff
@@ -43,38 +49,170 @@ const Login = () => {
     }
 
     const loginUser = (username, password) => {
-        authService.authUser(username, password);
+        authService.authUser(username, password)
+            .then(response => {
+                axios.get("http://" + backendServerURL + "/dashboard", {withCredentials: true} )
+                    .then(response => {
+                        if (response.status === 200) {
+                            navigate('/dashboard');
+                        }
+                    })
+                    .catch(error => {
+                        console.log(error.response);
+                    })
+            })
+        .catch(error => {
+            if (error.response === 401) {
+                setPasswordError('Invalid username or password');
+            }
+        });
+        }
+
+    const registerUser = (username, password, repeatPassword, email) => {
+        registrationService.createUser(username, password, repeatPassword, email)
+            .then(response => {
+                if (response.status === 200) {
+                    navigate('/dashboard');
+                }
+            })
+            .catch(error => {
+                error.response.data.map(fieldError => {
+                    if (fieldError.field === "username") {
+                        setUsernameError(fieldError.defaultMessage)
+                    }
+
+                    if (fieldError.field === "password") {
+                        setPasswordError(fieldError.defaultMessage)
+                    }
+
+                    if (fieldError.field === "repeatPassword") {
+                        setRepeatPasswordError(fieldError.defaultMessage)
+                    }
+                    
+                    if (fieldError.field === "email") {
+                        setEmailError(fieldError.defaultMessage)
+                    }
+                })
+            })
     }
 
+    const processUsernameInput = (event) => {
```

Comment:
Wszystkie nasze metody _processXXX_ robią dokładnie to samo, tylko dla innych pól. Wymyśliłem sposób na to, żeby to uwspólnić i nie powielać logiki. Wykorzystując programowanie funkcyjne można to napisać w ten sposób:

    const processUsernameInput = (event) => {
        const providedUsername = event.target.value;
        processFieldInput(providedUsername, validationService.validateUsername, setUsername, setUsernameError);
    }

    const processFieldInput = (fieldValue, validationMethod, setFieldValueMethod, setFieldErrorMethod) => {
        const errorMessage = validationMethod(fieldValue);
        if(errorMessage) {
            setFieldErrorMethod(errorMessage.message);
        } else  {
            setFieldErrorMethod('');
        }

        setFieldValueMethod(fieldValue);
    }


I następnie wszędzie re-użyć metody processFieldInput. W ten sposób nie duplikujemy logiki, zmniejszamy ilość kodu i mamy komponent potencjalnie łatwiejszy w testowaniu.

Reply:
Kurde, znakomite. Zmieniłam wszędzie, oprócz repeatPassword, bo tam do walidacji potrzebne są dwa parametry (a dla tego jednego inputa stwierdziłam, że nie warto pisać takiej generycznej metody).